v2/tests/git_initializer_test.go

```diff
@@ -179,46 +231,150 @@ func TestMain(t *testing.T) {
 
 			e := eList.Items[0]
 
-			logEntryCh, errCh, err :=
-				client.Core().Events().Logs().Stream(
-					ctx,
-					e.ID,
-					&core.LogsSelector{Container: "vcs"},
-					&core.LogStreamOptions{})
-			require.NoError(t, err, "error acquiring log stream")
-
-			for {
-				select {
-				case logEntry, ok := <-logEntryCh:
-					if ok {
-						log.Println(logEntry.Message)
-					} else {
-						logEntryCh = nil
-					}
-				case err, ok := <-errCh:
-					if ok {
-						t.Fatalf("error from log stream: %s", err)
-					}
-					errCh = nil
-				case <-ctx.Done():
-					break
-				}
-
-				// log and err channels empty, let's check worker status
-				if logEntryCh == nil && errCh == nil {
-					for {
-						event, err := client.Core().Events().Get(ctx, e.ID)
-						require.NoError(t, err, "error getting event")
-
-						phase := event.Worker.Status.Phase
-						if phase.IsTerminal() {
-							require.Equal(t, tc.terminalPhase, phase, "worker's terminal phase does not match expected")
-							break
-						}
-					}
-					break
-				}
+			assertWorkerTerminalPhase(t, ctx, client, e, tc)
+			assertWorkerLogs(t, ctx, client, e, tc)
+
+			// We expect the vcs init container to fail for this test,
+			// therefore, assertions on jobs are not applicable
+			if tc.name == "GitHub - vcs failure" {
+				return
 			}
+
+			assertJobTerminalPhase(t, ctx, client, e, tc)
+			assertJobLogs(t, ctx, client, e, tc)
 		})
 	}
 }
+
+func assertWorkerTerminalPhase(
+	t *testing.T,
+	ctx context.Context,
+	client sdk.APIClient,
+	e core.Event,
+	tc testcase) {
+
+	for {
+		event, err := client.Core().Events().Get(ctx, e.ID)
+		require.NoError(t, err, "error getting event")
+
+		phase := event.Worker.Status.Phase
+		if phase.IsTerminal() {
+			require.Equal(t, tc.expectedWorkerPhase, phase, "worker's terminal phase does not match expected")
+			break
+		}
+	}
```

Comment:
If a worker ever gets stuck in a non-terminal phase, this will loop infinitely.

If I may suggest, you can use `client.Core().Events().Workers().WatchStatus(ctx, e.ID)`. It will give you back two channels-- one for status updates and one for errors.

(Tangent, this is more efficient also because it only makes a single connection to the API server and the underlying goroutine sleeps when there aren't updates.)

With those channels in hand, you can do a select like this to implement a timeout so we never get stuck even if a worker gets stuck:

```
statucCh, errCh, err := client.Core().Events().Workers().WatchStatus(ctx, e.ID)
require.NoError(t, err)
timer := time.NewTimer( <some timeout duration> )
defer timer.Stop()
for {
  select {
    case status := <-statusCh:
      // If terming, assert expected phase
    case err := <- errCh:
      // Deal with error
    case <- timer.C:
      // Deal with timeout
  }
}
```

Reply:
This is much nicer; I had been sneakily using the context timeout as a lazyman's test timeout.  Updated with this suggestion.